src/object.d

```diff
@@ -2104,17 +2104,23 @@ auto byKeyValue(T : V[K], K, V)(T aa) pure nothrow @nogc
                 private void* keyp;
                 private void* valp;
 
-                @property ref key() inout { return *cast(substInout!K*)keyp; }
-                @property ref value() inout { return *cast(substInout!V*)valp; }
+                @property auto key() inout
+                {
+                    return () @trusted { return *cast(substInout!K*)keyp; } ();
+                };
+                @property auto value() inout
+                {
+                    return () @trusted { return *cast(substInout!V*)valp; } ();
+                };
```

Comment:
There is a problem here in that the AA can rehash the buckets, making any existing range point at invalid elements. I think in order to make this `@safe` we need to verify the range is still pointing at a valid element.

In particular the fetching of the value returns the pointer + key.sizeof, which even if the element pointer is null is going to be sketchy.

I still prefer to see this not being a `@trusted` escape, but instead mark the implementation as `@safe` and have the implementation use an escape.

We also have to be careful here of any kind of postblit in `K` or `V`. If you trust the whole thing, then it is going to inadvertently trust that part too.

Reply:
@schveiguy the "abandoned" elements will be still valid memory. All we need to make sure is they are cleared properly (by means of `move` in all likelihood). That would make old ranges iterate the wrong elements but not unsafe.

Reply:
The abandoned elements themselves are valid, and are set to null.

However, the code for fetching the value looks like this (from here: https://github.com/somzzz/druntime/blob/84af68f87d077d9632e7dc58da04670316d4219c/src/rt/aaA.d#L705):

```D
void* _aaRangeFrontValue(Range r)
    {
        return r.buckets[r.idx].entry + r.valoff;
    }
```

Here, `r.buckets` is really `r.impl.buckets` (`impl` points at the actual AA instance, so we are OK there). The `r.idx` is the range's local copy of which bucket it is currently looking at.

entry is possibly `null` (shouldn't be on a valid AA range, but if a rehash happens, it could be). So we are adding an arbitrary amount to a `null` pointer, and dereferencing that without considering it (unsafe).

Another consideration that I had figured out in the past but forgot to mention it here, it's very possible `r.idx >= r.impl.buckets.length` if the rehash has shrunk the array. Now we are definitely in unsafe territory.

All this can be fixed by returning `null` if `r.idx >= r.buckets.length` or the appropriate bucket pointer is `null` (just don't add the offset). And these aren't bounds-checked, since druntime is compiled in release mode.

Reply:
Looking at this more closely, it's really the generation of the `Pair` struct that is where the safety issues may lie.